elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryCancelThread.java

```diff
@@ -0,0 +1,112 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import javax.ws.rs.core.MultivaluedHashMap;
+import javax.ws.rs.core.MultivaluedMap;
+
+/**
+ * Runnable thread for cancelling AsyncQuery transactions
+ * beyond the max run time or if it has status CANCELLED.
+ */
+@Slf4j
+@Data
+@AllArgsConstructor
+public class AsyncQueryCancelThread implements Runnable {
+
+    private int maxRunTimeSeconds;
+    private Elide elide;
+    private AsyncQueryDAO asyncQueryDao;
+    private EntityDictionary dictionary;
+    private RSQLFilterDialect filterParser;
+
+    @Override
+    public void run() {
+        cancelAsyncQuery();
+    }
+
+    /**
+     * This method cancels queries based on threshold.
+     */
+    protected void cancelAsyncQuery() {
+
+        try {
+
+            TransactionRegistry transactionRegistry = elide.getTransactionRegistry();
+            Map<UUID, DataStoreTransaction> runningTransactionMap = transactionRegistry.getRunningTransactions();
+            String filterByStatus = "status=in=(" + QueryStatus.CANCELLED.toString() + ","
+                    + QueryStatus.PROCESSING.toString() + ","
+                    + QueryStatus.QUEUED.toString() + ")";
+            FilterExpression fltStatusExpression = filterParser.parseFilterExpression(filterByStatus,
+                    AsyncQuery.class, false);
+            Collection<AsyncQuery> asyncQueryCollection =
+                    asyncQueryDao.getActiveAsyncQueryCollection(fltStatusExpression);
+
+            Set<UUID> runningTransactions = runningTransactionMap.keySet();
+
+            Set<UUID> asyncTransactions = asyncQueryCollection.stream()
+                    .filter(query -> query.getStatus() == QueryStatus.CANCELLED
+                    || TimeUnit.SECONDS.convert(Math.abs(query.getUpdatedOn().getTime()
+                    - new Date(System.currentTimeMillis()).getTime()), TimeUnit.MILLISECONDS) > maxRunTimeSeconds)
+                    .map(AsyncQuery::getRequestId)
+            .collect(Collectors.toSet());
+
+            Set<UUID> queriesToCancel = Sets.intersection(runningTransactions, asyncTransactions);
+
+            List<String> asyncQueryList = new ArrayList<String>();
+            queriesToCancel.stream()
+               .forEach((tx) -> {
+                   JsonApiDocument jsonApiDoc = new JsonApiDocument();
+                   MultivaluedMap<String, String> queryParams = new MultivaluedHashMap<String, String>();
+                   RequestScope scope = new RequestScope("query", NO_VERSION, jsonApiDoc,
+                           transactionRegistry.getRunningTransaction(tx), null, queryParams,
+                           tx, elide.getElideSettings());
+                   transactionRegistry.getRunningTransaction(tx).cancel(scope);
+                   List<String> cancelledQuery = asyncQueryCollection.stream().filter(query
```

Comment:
A couple of improvements on this code.  I think we should rename some of the local variables for clarity.  When dealing with ids, we shouldn't call them queries or transactions but rather identifiers - so we can distinguish the collections from each other.

Second, I'm not sure how much of an improvement this is, but I would probably have written the mapping from UUIDs to kill to queries to kill as follows:

```java
Set<Query> queriesToCancel = queryIdsToCancel.stream()
        .map(uuid -> { 
            asyncQueryCollection.stream()
               .filter(query -> query.getRequestId().equals(uuid))
               .findFirst().orElseThrow(IllegalStateException::new)
         })
         .collect(Collectors.toSet());
```

Because naming is hard, it is sometimes more readable to collapse the mapping into a single logical block without intermediate state.  Alternatively, we could write a separate function to do the same thing.